### PR TITLE
Test suite specific loader

### PR DESCRIPTION
**Pull request by richtja**
We have multiple test suites inside one Job and each of them can have different
configurations from the user. Unfortunately, we have just one instance of the
test loader which is not able to deal with different configurations of test
suites. This commit creates for each test suite instance its own test loader
which will work with the right configuration.

Signed-off-by: Jan Richter <jarichte@redhat.com>

### CODE REVIEW

**Review by pevogam — COMMENTED** [review]
This is an awesome idea @richtja! I guess this will need some elaborate testing but it really gives some more sense to the multiple test suite architecture.

**Comment by beraldoleal** [comment]
Hi @richtja this is nice, super cool.

But, please help me with something: I'm might be wrong, but I had understood from our conversations that after LTS, all the legacy runner and loader stuff would be removed. Probably I'm missing something.

**Comment by pevogam** [comment]
Hi @beraldoleal,

> But, please help me with something: I'm might be wrong, but I had understood from our conversations that after LTS, all the legacy runner and loader stuff would be removed. Probably I'm missing something.

I still it is still nice if this could be added and at least backported for the serial runner as it is still used by many people that could benefit from multiple test suites with nonhomogeneous runners. Also, I am still not sure why the presence of a parallel runner should necessarily imply the serial runner is obsolete but I know I am digging into something that was already decided.

**Comment by beraldoleal** [comment]
> Hi @beraldoleal,
> 
> > But, please help me with something: I'm might be wrong, but I had understood from our conversations that after LTS, all the legacy runner and loader stuff would be removed. Probably I'm missing something.
> 
> I still it is still nice if this could be added and at least backported for the serial runner as it is still used by many people that could benefit from multiple test suites with nonhomogeneous runners. Also, I am still not sure why the presence of a parallel runner should necessarily imply the serial runner is obsolete but I know I am digging into something that was already decided.

My only concern is that this could have some impact on the decision to migrate to the new runner. Users will see a new "feature" on the legacy and will have the tendency to postpone the migration. But again, I'm just pointing out, and hopefully I'm not create to much noise here.

If it is really important and it is a bug, and IIUC, then yes, I agree with your @pevogam, probably we will need to backport this.

**Comment by pevogam** [comment]
> > I still it is still nice if this could be added and at least backported for the serial runner as it is still used by many people that could benefit from multiple test suites with nonhomogeneous runners. Also, I am still not sure why the presence of a parallel runner should necessarily imply the serial runner is obsolete but I know I am digging into something that was already decided.
> 
> My only concern is that this could have some impact on the decision to migrate to the new runner. Users will see a new "feature" on the legacy and will have the tendency to postpone the migration. But again, I'm just pointing out, and hopefully I'm not create to much noise here.
> 
> If it is really important and it is a bug, and IIUC, then yes, I agree with your @pevogam, probably we will need to backport this.

Actually yes, you have more right here since the main point of backporting to the LTS is for stability reasons and fixes only, no new features.

So I guess it boils down to the question whether parallelization should be enforced for everyone and thus they should be "stimulated" to move away from older versions or not. Forgive me if I forgot something critical but was there a way to use the parallel runner for serial test running tasks?

Of course, we could move this discussion to another place since I don't want to kidnap the PR's main topic here and my last question got slightly off-topic.

**Comment by willianrampazzo** [comment]
>Forgive me if I forgot something critical but was there a way to use the parallel runner for serial test running tasks?
> 

Hey @pevogam, not joining the discussion, but just giving a pointer here. Check the following option:

```
435:nrunner.max_parallel_tasks
436-~~~~~~~~~~~~~~~~~~~~~~~~~~
437:Number of maximum number tasks running in parallel. You can disable
438:parallel execution by setting this to 1. Defaults to the amount of
439-CPUs on this machine.
440-
441-* Default: 8
442-* Type: <class 'int'>
```

Command-line parameter:

```
  --nrunner-max-parallel-tasks NRUNNER.MAX_PARALLEL_TASKS
                        Number of maximum number tasks running in parallel. You can disable parallel execution by
                        setting this to 1. Defaults to the amount of CPUs on this machine.
```

**Comment by richtja** [comment]
> > Hi @beraldoleal,
> > > But, please help me with something: I'm might be wrong, but I had understood from our conversations that after LTS, all the legacy runner and loader stuff would be removed. Probably I'm missing something.
> > 
> > 
> > I still it is still nice if this could be added and at least backported for the serial runner as it is still used by many people that could benefit from multiple test suites with nonhomogeneous runners. Also, I am still not sure why the presence of a parallel runner should necessarily imply the serial runner is obsolete but I know I am digging into something that was already decided.
> 
> My only concern is that this could have some impact on the decision to migrate to the new runner. Users will see a new "feature" on the legacy and will have the tendency to postpone the migration. But again, I'm just pointing out, and hopefully I'm not create to much noise here.

@beraldoleal Thanks for your comments.

I created this because I wasn't able to run multiple external runners on the legacy runner in one job and I am still using legacy runner because AFAIK the nrunner don't support avocado-vt tests which I also want to run in the same job. I don't see this as a new feature, I see it as fixing the expected behavior. But you are right that I didn't consider that this can postpone the migration.

Actually we still didn't decide how the compatibility of avocado and avocado-vt will look like after we remove the legacy runner.

**Comment by beraldoleal** [comment]
Maybe we could focus on make avocado-vt running on the next runner. I'm not sure. @clebergnu any comments here?

**Comment by pevogam** [comment]
Hi @willianrampazzo,

> Command-line parameter:
> 
> ```
>   --nrunner-max-parallel-tasks NRUNNER.MAX_PARALLEL_TASKS
>                         Number of maximum number tasks running in parallel. You can disable parallel execution by
>                         setting this to 1. Defaults to the amount of CPUs on this machine.
> ```

Pointer more than welcome. However, then I guess my perception here was wrong since I was implicitly associating tasks with tests according to the implementations I saw previously. I guess this simply reuses the name "task" for a different purpose? Or perhaps we have a handler that could split N tests into M tasks where N!=M?

**Comment by willianrampazzo** [comment]
> Hi @willianrampazzo,
> 
> > Command-line parameter:
> > ```
> >   --nrunner-max-parallel-tasks NRUNNER.MAX_PARALLEL_TASKS
> >                         Number of maximum number tasks running in parallel. You can disable parallel execution by
> >                         setting this to 1. Defaults to the amount of CPUs on this machine.
> > ```
> 
> Pointer more than welcome. However, then I guess my perception here was wrong since I was implicitly associating tasks with tests according to the implementations I saw previously. I guess this simply reuses the name "task" for a different purpose? Or perhaps we have a handler that could split N tests into M tasks where N!=M?

If you look at BP003 (https://avocado-framework.readthedocs.io/en/82.0/blueprints/BP003.html) about the Task state machine, you will notice that it uses as an example one reference resolution.

Looking at the current implementation of a Task (https://github.com/avocado-framework/avocado/blob/master/avocado/core/nrunner.py#L617), it "wraps the execution of a runnable". The only use we have right now for a Task is to run a runnable from a reference resolution, but this is subject to change soon when we start to create runnable for the requirements and wrap them into Tasks as we discussed in the Requirements Resolver PR.

In a nutshell, a Task is a much broader entity than a simple test representation.

**Comment by richtja** [comment]
I am closing this, because our priority is to focus on nrunner where this problem doesn't exist. There is still problem of compatibility of nrunner and avocado-vt and @clebergnu started working on that at [avocado-vt  side](https://github.com/avocado-framework/avocado-vt/pull/2904)
